sql/inspect: support AS OF SYSTEM TIME for historical descriptor access

spilchen opened the pull request:
Prior to this change, INSPECT could fail late during execution when running AS OF SYSTEM TIME queries that referenced objects which did not exist at the specified timestamp. For example, querying a database that had not yet been created would trigger an internal error instead of a clear user facing message.

This change updates INSPECT’s planning logic to detect such cases earlier and return a more informative error (e.g. “database movr does not exist”). It does this by using a fixed timestamp and avoiding leased descriptors for ASOF queries.

Informs: #155483
Epic: CRDB-55075
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155974)
<!-- Reviewable:end -->

fqazi reviewed (changes requested):
> @fqazi reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155974)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/resolver.go` line 1244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155974#-OcICOZw4CFQLG58eg4r:-OcICOZxD736_fmNRN69:b25w0xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97c62b33a659dd6a081ddb6e3f9579952016f3d/pkg/sql/resolver.go#L1244)):*
> ```Go
> 	// the descriptor cache to ensure we get the correct historical value.
> 	if p.extendedEvalCtx.AsOfSystemTime != nil {
> 		p.runWithOptions(resolveFlags{skipCache: true}, resolve)
> ```

This shouldn't be needed if your doing a historical query we should resolve the correct timestamp. Unless it's really close to the current timestamp and even then at most you get V-1 in that case. The lease manager has historical query support.

Even then it should be readable for the data.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: support AS OF SYSTEM TIME for historical descriptor access

Prior to this change, INSPECT could fail late during execution when
running AS OF SYSTEM TIME queries that referenced objects which did not
exist at the specified timestamp. For example, querying a database that
had not yet been created would trigger an internal error instead of a
clear user facing message.

This change updates INSPECT’s planning logic to detect such cases
earlier and return a more informative error (e.g. “database movr does
not exist”). It does this by using a fixed timestamp and avoiding leased
descriptors for ASOF queries.

Informs: #155483
Epic: CRDB-55075
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155974)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/resolver.go` line 1244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155974#-OcICOZw4CFQLG58eg4r:-OcKzx542Jxi3WH-eEWr:bwzm0hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e97c62b33a659dd6a081ddb6e3f9579952016f3d/pkg/sql/resolver.go#L1244)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This shouldn't be needed if your doing a historical query we should resolve the correct timestamp. Unless it's really close to the current timestamp and even then at most you get V-1 in that case. The lease manager has historical query support.

Even then it should be readable for the data.
</blockquote></details>

Thanks. I think I misunderstood historical descriptors and the cache. I thought the cache had to be avoided. I removed this code (and another spot where I skipped the cache).


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @fqazi reviewed 1 of 5 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155974#-:-OcL84ZE2aDQkDGO5ZmY:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155974#-OcL8DY01xf70b6QCvuv:-OcL8DY01xf70b6QCvuw:bpj49eq):*
Nit: Probably need to update this message a bit.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
